### PR TITLE
feat(qlitics): qlitics for story-element-view, story-element action

### PR DESCRIPTION
**Pull request by Jeevan-Kishore**
Qlitics events are to be fired for every story element that is scrolled into view.

### CODE REVIEW

File: src/components/story-element.js [hunk 1]
```diff
@@ -94,12 +94,65 @@ const DEFAULT_TEMPLATES = {
...
+  componentDidMount() {
+    this.initiateObserver();
+  }
+
+  componentWillUnmount() {
+    this.destroyObserver();
+  }
+
+  initiateObserver = () => {
+    if(this.props.disableAnalytics === true || (global && !global.qlitics)) return false;
```
Comment (sinujohn91) [inline src/components/story-element.js]: what if qlitics loads after the component has mounted?

Comment (gja) [reply src/components/story-element.js]: good point. I think you'll need to do the global.qlitics = global.qlitics || [];

qlitics.push()

File: src/components/story-element.js [hunk 2]
```diff
@@ -94,12 +94,65 @@ const DEFAULT_TEMPLATES = {
...
   template() {
     const storyElement = this.props.element;
     const templates = Object.assign({}, DEFAULT_TEMPLATES, this.props.templates);
     return templates[storyElement.subtype] || templates[storyElement.type] || "div";
   }
 
+
+  emitElementQlitics() {
+    const {story = {}, card = {}, element = {}} = this.props;
+    global.qlitics('track', 'story-element-view', {
```
Comment (gja) [inline src/components/story-element.js]: use the qlitics.push()

File: src/components/story-element.js [hunk 3]
```diff
@@ -94,12 +94,65 @@ const DEFAULT_TEMPLATES = {
...
+
+  initiateObserver = () => {
+    if(this.props.disableAnalytics === true || (global && !global.qlitics)) return false;
+
+    const options = {
+      root: null,
+      rootMargin: '0px',
+      threshold: 1.0
+    };
+    this.observer = new IntersectionObserver(this.observerCallback, options);
```
Comment (gja) [inline src/components/story-element.js]: does having an individual IntersectionObserver have a perf impact? I have no idea

File: src/components/story-element.js (hunk 1 above)
Comment (Jeevan-Kishore) [reply src/components/story-element.js]: Added.

File: src/components/story-element.js (hunk 3 above)
Comment (Jeevan-Kishore) [reply src/components/story-element.js]: Taking a call post discussion to keep it for now

* TODO : refactor later

File: src/components/story-element.js [hunk 4]
```diff
@@ -94,12 +95,63 @@ const DEFAULT_TEMPLATES = {
...
+
+  destroyObserver = () => {
+    this.observer && this.observer.disconnect();
+  };
+
+  observerCallback = entries => {
+    entries.forEach(entry => {
+      if (entry.isIntersecting) {
+        this.emitElementQlitics();
+        this.destroyObserver(); //destroy observer once the call is made
```
Comment (sinujohn91) [inline src/components/story-element.js]: i think we need to fire it everytime it comes in view.

File: src/components/story-element.js [hunk 5]
```diff
@@ -94,12 +95,63 @@ const DEFAULT_TEMPLATES = {
...
     return templates[storyElement.subtype] || templates[storyElement.type] || "div";
   }
 
+
+  emitElementQlitics() {
+    const {story = {}, card = {}, element = {}} = this.props;
+    if(global.qlitics){
+      global.qlitics('track', 'story-element-view', getQliticsSchema(story,card,element));
+    } else {
+      global.qlitics=global.qlitics||function(){(qlitics.q=qlitics.q||[]).push(arguments);};
```
Comment (sinujohn91) [inline src/components/story-element.js]: can you explain what you are trying to do here?

Comment (Jeevan-Kishore) [reply src/components/story-element.js]: @gja recommended checking the earlier implementations, the code ripped from [this](https://github.com/quintype/quintype-node-framework/blob/0dbf683ea59547d5c7f71bfe9467a8c3c1762d3b/client/analytics.js#L6)

Reading through the code, If qlitics is not found, it gets allocated to a function which takes in arguments to push into qlitics.q

File: src/components/story-element.js (hunk 4 above)
Comment (Jeevan-Kishore) [reply src/components/story-element.js]: done

**Pull request merged**